Plugin doesn't work with Android Gradle Plugin 4.1.0 alpha 01

With new AGP 4.1.0-alpha01, ExtensionKt.testTastOf method throws NoSuchMethodError exception.

```
* Exception is:
org.gradle.api.ProjectConfigurationException: A problem occurred configuring project ':common-extension'.
	at org.gradle.configuration.project.LifecycleProjectEvaluator.wrapException(LifecycleProjectEvaluator.java:80)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.addConfigurationFailure(LifecycleProjectEvaluator.java:73)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.access$600(LifecycleProjectEvaluator.java:53)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate.run(LifecycleProjectEvaluator.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject$1.run(LifecycleProjectEvaluator.java:112)
	at org.gradle.internal.Factories$1.create(Factories.java:26)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:189)
	at org.gradle.internal.work.StopShieldingWorkerLeaseService.withLocks(StopShieldingWorkerLeaseService.java:40)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.withProjectLock(DefaultProjectStateRegistry.java:271)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.withMutableState(DefaultProjectStateRegistry.java:265)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.withMutableState(DefaultProjectStateRegistry.java:226)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$EvaluateProject.run(LifecycleProjectEvaluator.java:96)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:68)
	at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:700)
	at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:142)
	at org.gradle.execution.TaskPathProjectEvaluator.configure(TaskPathProjectEvaluator.java:36)
	at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:64)
	at org.gradle.configuration.DefaultProjectsPreparer.prepareProjects(DefaultProjectsPreparer.java:61)
	at org.gradle.configuration.BuildOperatingFiringProjectsPreparer$ConfigureBuild.run(BuildOperatingFiringProjectsPreparer.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.configuration.BuildOperatingFiringProjectsPreparer.prepareProjects(BuildOperatingFiringProjectsPreparer.java:40)
	at org.gradle.initialization.DefaultGradleLauncher.prepareProjects(DefaultGradleLauncher.java:218)
	at org.gradle.initialization.DefaultGradleLauncher.doClassicBuildStages(DefaultGradleLauncher.java:155)
	at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:140)
	at org.gradle.initialization.DefaultGradleLauncher.getConfiguredBuild(DefaultGradleLauncher.java:114)
	at org.gradle.internal.invocation.GradleBuildController$1.create(GradleBuildController.java:74)
	at org.gradle.internal.invocation.GradleBuildController$1.create(GradleBuildController.java:67)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:189)
	at org.gradle.internal.work.StopShieldingWorkerLeaseService.withLocks(StopShieldingWorkerLeaseService.java:40)
	at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:67)
	at org.gradle.internal.invocation.GradleBuildController.configure(GradleBuildController.java:61)
	at org.gradle.tooling.internal.provider.runner.ClientProvidedPhasedActionRunner.run(ClientProvidedPhasedActionRunner.java:62)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:63)
	at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
	at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:39)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:51)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:45)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:45)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:50)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:47)
	at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:80)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:47)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:31)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:42)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:28)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:78)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:52)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:60)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:38)
	at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:68)
	at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:38)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:37)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:26)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:43)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:29)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:60)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:32)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:55)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:41)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:48)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:32)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:68)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
	at org.gradle.util.Swapper.swap(Swapper.java:38)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:82)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.internal.event.ListenerNotificationException: Failed to notify project evaluation listener.
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:86)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:325)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:235)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy35.afterEvaluate(Unknown Source)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:191)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:188)
	at org.gradle.api.internal.project.DefaultProject.stepEvaluationListener(DefaultProject.java:1438)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate.run(LifecycleProjectEvaluator.java:197)
	... 113 more
Caused by: java.lang.NoSuchMethodError: com.android.build.gradle.internal.variant.BaseVariantData.getScope()Lcom/android/build/gradle/internal/scope/VariantScope;
	at de.mannodermaus.gradle.plugins.junit5.internal.ExtensionsKt.testTaskOf(Extensions.kt:208)
	at de.mannodermaus.gradle.plugins.junit5.AndroidJUnitPlatformPlugin$configureUnitTests$1.execute(Plugin.kt:61)
	at de.mannodermaus.gradle.plugins.junit5.AndroidJUnitPlatformPlugin$configureUnitTests$1.execute(Plugin.kt:19)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1$1.run(DefaultCollectionCallbackActionDecorator.java:100)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.reapply(DefaultUserCodeApplicationContext.java:60)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:97)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:94)
	at org.gradle.api.internal.DefaultDomainObjectCollection.all(DefaultDomainObjectCollection.java:163)
	at de.mannodermaus.gradle.plugins.junit5.AndroidJUnitPlatformPlugin.configureUnitTests(Plugin.kt:60)
	at de.mannodermaus.gradle.plugins.junit5.AndroidJUnitPlatformPlugin.access$configureUnitTests(Plugin.kt:19)
	at de.mannodermaus.gradle.plugins.junit5.AndroidJUnitPlatformPlugin$apply$3.execute(Plugin.kt:41)
	at de.mannodermaus.gradle.plugins.junit5.AndroidJUnitPlatformPlugin$apply$3.execute(Plugin.kt:19)
	at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingAction$1.lambda$run$0(DefaultListenerBuildOperationDecorator.java:152)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.reapply(DefaultUserCodeApplicationContext.java:60)
	at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingAction$1.run(DefaultListenerBuildOperationDecorator.java:152)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingAction.execute(DefaultListenerBuildOperationDecorator.java:149)
	at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:92)
	at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:80)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:231)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:150)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	... 123 more
```

--- a/buildSrc/src/main/kotlin/Artifacts.kt
+++ b/buildSrc/src/main/kotlin/Artifacts.kt
@@ -35,7 +35,7 @@ object Artifacts {
       platform = Java,
       groupId = "de.mannodermaus.gradle.plugins",
       artifactId = "android-junit5",
-      currentVersion = "1.6.1.0-SNAPSHOT",
+      currentVersion = "1.6.0.1-SNAPSHOT",
       latestStableVersion = "1.6.0.0",
       license = license,
       description = "Unit Testing with JUnit 5 for Android."

--- a/buildSrc/src/main/kotlin/Libs.kt
+++ b/buildSrc/src/main/kotlin/Libs.kt
@@ -40,6 +40,8 @@ object Libs {
         Versions.com_android_tools_build_gradle_36x
     const val com_android_tools_build_gradle_40x: String = "com.android.tools.build:gradle:" +
         Versions.com_android_tools_build_gradle_40x
+    const val com_android_tools_build_gradle_41x: String = "com.android.tools.build:gradle:" +
+        Versions.com_android_tools_build_gradle_41x
 
     /**
      * https://developer.android.com/studio */

--- a/buildSrc/src/main/kotlin/Versions.kt
+++ b/buildSrc/src/main/kotlin/Versions.kt
@@ -18,6 +18,7 @@ object Versions {
   const val com_android_tools_build_gradle_35x: String = "3.5.0"
   const val com_android_tools_build_gradle_36x: String = "3.6.0"
   const val com_android_tools_build_gradle_40x: String = "4.0.0-beta01"
+  const val com_android_tools_build_gradle_41x: String = "4.1.0-alpha01"
   const val com_android_tools_build_gradle: String = com_android_tools_build_gradle_35x
 
   const val lint_gradle: String = "26.5.0"

--- a/plugin/CHANGELOG.md
+++ b/plugin/CHANGELOG.md
@@ -4,7 +4,9 @@ Change Log
 ## Unreleased
 
 #### Added
+- Compatibility with Android Gradle Plugin 4.1.x
 #### Changed
+- Restricted visibility of internal APIs to prevent them from leaking into consumer code
 #### Fixed
 #### Removed
 

--- a/plugin/android-junit5/build.gradle.kts
+++ b/plugin/android-junit5/build.gradle.kts
@@ -50,6 +50,7 @@ gradlePlugin {
 // Use JUnit 5
 tasks.withType<Test> {
   useJUnitPlatform()
+  failFast = true
   testLogging {
     events = setOf(TestLogEvent.STARTED, TestLogEvent.SKIPPED, TestLogEvent.FAILED)
     exceptionFormat = TestExceptionFormat.FULL
@@ -67,6 +68,7 @@ tasks.named("processTestResources", Copy::class.java).configure {
       "AGP_35X" to Versions.com_android_tools_build_gradle_35x,
       "AGP_36X" to Versions.com_android_tools_build_gradle_36x,
       "AGP_40X" to Versions.com_android_tools_build_gradle_40x,
+      "AGP_41X" to Versions.com_android_tools_build_gradle_41x,
       "KOTLIN" to Versions.org_jetbrains_kotlin,
 
       "KOTLIN_STD_LIB" to Libs.kotlin_stdlib,
@@ -112,7 +114,8 @@ data class AgpConfiguration(val version: String, val dependency: String) {
 private val agpConfigurations = listOf(
     AgpConfiguration("3.5", Libs.com_android_tools_build_gradle_35x),
     AgpConfiguration("3.6", Libs.com_android_tools_build_gradle_36x),
-    AgpConfiguration("4.0", Libs.com_android_tools_build_gradle_40x)
+    AgpConfiguration("4.0", Libs.com_android_tools_build_gradle_40x),
+    AgpConfiguration("4.1", Libs.com_android_tools_build_gradle_41x)
 )
 
 configurations {

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/Interop.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/Interop.kt
@@ -1,7 +1,6 @@
 package de.mannodermaus.gradle.plugins.junit5
 
 import com.android.build.gradle.api.BaseVariant
-import com.android.build.gradle.internal.variant.BaseVariantData
 import de.mannodermaus.gradle.plugins.junit5.internal.invokeTyped
 import de.mannodermaus.gradle.plugins.junit5.internal.reflectiveMethod
 import org.gradle.api.Project
@@ -22,17 +21,6 @@ import java.io.File
  * BaseVariant
  * =====================================================================================================================
  */
-
-/**
- * Obtains the VariantData of the provided Variant.
- *
- * @because 'BaseVariant.variantData' is protected and can't be accessed through Kotlin directly
- * @return The variant data
- */
-val BaseVariant.variantData: BaseVariantData
-  get() = this.reflectiveMethod("getVariantData")
-      ?.invokeTyped<BaseVariantData>(this)
-      ?: throw IllegalArgumentException("BaseVariant.variantData not available")
 
 /**
  * Obtains the Java class directory for the provided variant in a safe manner.

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/ConfigurationKind.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/ConfigurationKind.kt
@@ -1,6 +1,6 @@
 package de.mannodermaus.gradle.plugins.junit5.internal
 
-enum class ConfigurationKind(internal val value: String) {
+internal enum class ConfigurationKind(internal val value: String) {
   APP(""),
   TEST("test"),
   ANDROID_TEST("androidTest")

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/ConfigurationScope.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/ConfigurationScope.kt
@@ -1,6 +1,6 @@
 package de.mannodermaus.gradle.plugins.junit5.internal
 
-enum class ConfigurationScope(internal vararg val values: String) {
+internal enum class ConfigurationScope(internal vararg val values: String) {
   API("api", "compile"),
   IMPLEMENTATION("implementation", "compile"),
   COMPILE_ONLY("compileOnly", "provided"),

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/Extensions.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/Extensions.kt
@@ -202,11 +202,33 @@ val BaseVariant.instrumentationTestVariant: TestVariant?
   }
 
 /**
+ * Composes the name of the variant-specific Gradle task with the given [prefix] and [suffix].
+ * For example, to obtain the "free" variant's assemble task name, call `variant.getTaskName(prefix = "assemble")`.
+ */
+internal fun BaseVariant.getTaskName(prefix: String = "", suffix: String = ""): String {
+  // At least one value must be provided
+  require(prefix.isNotEmpty() || suffix.isNotEmpty())
+
+  return StringBuilder().apply {
+    append(prefix)
+    append(if (isEmpty()) {
+      name
+    } else {
+      name.capitalize()
+    })
+    append(suffix.capitalize())
+  }.toString()
+}
+
+/**
  * Obtains the {AndroidUnitTest} for the provided variant.
  */
-fun TaskContainer.testTaskOf(variant: BaseVariant): AndroidUnitTest {
-  val taskName = variant.variantData.scope.getTaskName(VariantTypeCompat.UNIT_TEST_PREFIX,
-      VariantTypeCompat.UNIT_TEST_SUFFIX)
+internal fun TaskContainer.testTaskOf(variant: BaseVariant): AndroidUnitTest {
+  // From AGP 4.1 onwards, there is no Scope API on VariantData anymore.
+  // Task names must be constructed manually
+  val taskName = variant.getTaskName(
+      prefix = VariantTypeCompat.UNIT_TEST_PREFIX,
+      suffix = VariantTypeCompat.UNIT_TEST_SUFFIX)
   return getByName(taskName) as AndroidUnitTest
 }
 

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/Extensions.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/Extensions.kt
@@ -10,14 +10,10 @@ import com.github.zafarkhaja.semver.Version
 import de.mannodermaus.gradle.plugins.junit5.AndroidJUnitPlatformPlugin
 import de.mannodermaus.gradle.plugins.junit5.JUnit5TaskConfig
 import de.mannodermaus.gradle.plugins.junit5.VariantTypeCompat
-import de.mannodermaus.gradle.plugins.junit5.internal.ConfigurationKind.APP
-import de.mannodermaus.gradle.plugins.junit5.variantData
 import org.gradle.api.GradleException
 import org.gradle.api.Project
 import org.gradle.api.Task
-import org.gradle.api.artifacts.Configuration
 import org.gradle.api.artifacts.dsl.DependencyHandler
-import org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestFramework
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.LogLevel.ERROR
 import org.gradle.api.logging.LogLevel.INFO
@@ -26,8 +22,6 @@ import org.gradle.api.logging.Logger
 import org.gradle.api.plugins.ExtensionAware
 import org.gradle.api.plugins.ExtraPropertiesExtension
 import org.gradle.api.tasks.TaskContainer
-import org.gradle.api.tasks.testing.Test
-import org.gradle.api.tasks.testing.junitplatform.JUnitPlatformOptions
 import org.gradle.util.GradleVersion
 import java.util.Properties
 
@@ -45,7 +39,7 @@ internal fun requireGradle(version: String, message: () -> String) {
   }
 }
 
-fun requireVersion(actual: String, required: String, message: () -> String) {
+internal fun requireVersion(actual: String, required: String, message: () -> String) {
   val actualVersion = Version.valueOf(actual)
   val requiredVersion = Version.valueOf(required)
   require(actualVersion.greaterThanOrEqualTo(requiredVersion)) {
@@ -65,7 +59,7 @@ internal fun loadProperties(resource: String): Properties {
  * and splits it into a list using the delimiter.
  * Returns an empty List if the key doesn't exist in the Map.
  */
-fun Map<String, String>.getAsList(key: String, delimiter: String = ","): List<String> =
+internal fun Map<String, String>.getAsList(key: String, delimiter: String = ","): List<String> =
     this[key]?.split(delimiter) ?: emptyList()
 
 /**
@@ -74,7 +68,7 @@ fun Map<String, String>.getAsList(key: String, delimiter: String = ","): List<St
  * the value is appended to the end of the current value
  * using the given delimiter.
  */
-fun MutableMap<String, String>.append(
+internal fun MutableMap<String, String>.append(
     key: String, value: String, delimiter: String = ","): String? {
   val insertedValue = if (containsKey(key)) {
     "${this[key]}$delimiter$value"
@@ -91,7 +85,7 @@ fun MutableMap<String, String>.append(
  * Log a message with the Android Gradle Plugin style syntax,
  * which will cause Android Studio to pick it up & display it inside the Messages window.
  */
-fun Logger.agpLog(level: LogLevel, message: String) {
+internal fun Logger.agpLog(level: LogLevel, message: String) {
   val pair: Pair<String, (String) -> Unit> = when (level) {
     ERROR -> "error" to { s -> error(s) }
     WARN -> "warning" to { s -> warn(s) }
@@ -107,7 +101,7 @@ fun Logger.agpLog(level: LogLevel, message: String) {
 /**
  * Create & add an Extension to the given container by name.
  */
-inline fun <reified T> Any.extend(
+internal inline fun <reified T> Any.extend(
     name: String,
     args: Array<Any> = emptyArray(),
     noinline init: ((T) -> Unit)? = null): T {
@@ -119,11 +113,11 @@ inline fun <reified T> Any.extend(
 
   // Create & Configure the new extension
   val created: T = this.extensions.create(name, T::class.java, *args)
-  init?.let { init(created) }
+  init?.let { it(created) }
   return created
 }
 
-fun Any.extensionExists(name: String): Boolean {
+internal fun Any.extensionExists(name: String): Boolean {
   if (this !is ExtensionAware) {
     throw IllegalArgumentException("Argument is not ExtensionAware: $this")
   }
@@ -135,7 +129,7 @@ fun Any.extensionExists(name: String): Boolean {
  * Obtain an Extension by name & directly cast it to the expected type.
  */
 @Suppress("UNCHECKED_CAST")
-fun <T> Any.extensionByName(name: String): T {
+internal fun <T> Any.extensionByName(name: String): T {
   if (this !is ExtensionAware) {
     throw IllegalArgumentException("Argument is not ExtensionAware: $this")
   }
@@ -146,27 +140,27 @@ fun <T> Any.extensionByName(name: String): T {
 /**
  * Log the provided info message using the plugin's Log Tag.
  */
-fun Logger.junit5Info(text: String) {
+internal fun Logger.junit5Info(text: String) {
   info("[android-junit5]: $text")
 }
 
 /**
  * Log the provided warning message using the plugin's Log Tag.
  */
-fun Logger.junit5Warn(text: String) {
+internal fun Logger.junit5Warn(text: String) {
   warn("[android-junit5]: $text")
 }
 
 /**
  * Shorthand function to check for the existence of a plugin on a Project.
  */
-fun Project.hasPlugin(name: String) = this.plugins.findPlugin(name) != null
+internal fun Project.hasPlugin(name: String) = this.plugins.findPlugin(name) != null
 
 /**
  * Access the Android extension applied by a respective plugin.
  * Equivalent to "Project#android" in Groovy.
  */
-val Project.android: BaseExtension
+internal val Project.android: BaseExtension
   get() = this.extensions.getByName("android") as BaseExtension
 
 internal fun Project.junit5ConfigurationOf(variant: BaseVariant) =
@@ -176,14 +170,14 @@ internal fun Project.junit5ConfigurationOf(variant: BaseVariant) =
  * Access the extra properties of a DependencyHandler.
  * Equivalent to "DependencyHandler#ext" in Groovy.
  */
-val DependencyHandler.ext: ExtraPropertiesExtension
+internal val DependencyHandler.ext: ExtraPropertiesExtension
   get() {
     val aware = this as ExtensionAware
     return aware.extensions.getByName(
         ExtraPropertiesExtension.EXTENSION_NAME) as ExtraPropertiesExtension
   }
 
-val BaseVariant.unitTestVariant: UnitTestVariant
+internal val BaseVariant.unitTestVariant: UnitTestVariant
   get() {
     if (this !is TestedVariant) {
       throw IllegalArgumentException("Argument is not TestedVariant: $this")
@@ -192,7 +186,7 @@ val BaseVariant.unitTestVariant: UnitTestVariant
     return this.unitTestVariant
   }
 
-val BaseVariant.instrumentationTestVariant: TestVariant?
+internal val BaseVariant.instrumentationTestVariant: TestVariant?
   get() {
     if (this !is TestedVariant) {
       throw IllegalArgumentException("Argument is not TestedVariant: $this")
@@ -238,7 +232,7 @@ internal fun TaskContainer.testTaskOf(variant: BaseVariant): AndroidUnitTest {
  * in which case the existing task is returned.
  */
 @Suppress("UNCHECKED_CAST")
-fun TaskContainer.maybeCreate(name: String, group: String? = null): Task {
+internal fun TaskContainer.maybeCreate(name: String, group: String? = null): Task {
   val existing = findByName(name)
   return if (existing != null) {
     existing

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/Reflective.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/internal/Reflective.kt
@@ -13,7 +13,7 @@ import java.lang.reflect.Method
  * @param paramTypes Parameter types of the method's signature
  * @return A reference to the method with that signature, or null if it doesn't exist
  */
-fun Any.reflectiveMethod(named: String, vararg paramTypes: Class<*>): Method? =
+internal fun Any.reflectiveMethod(named: String, vararg paramTypes: Class<*>): Method? =
     try {
       val targetClass = if (this is Class<*>) {
         this
@@ -34,6 +34,6 @@ fun Any.reflectiveMethod(named: String, vararg paramTypes: Class<*>): Method? =
  * @param params Parameters to call the method with
  * @return Result, cast to the desired type
  */
-fun <T> Method.invokeTyped(target: Any? = null, vararg params: Any?): T? {
+internal fun <T> Method.invokeTyped(target: Any? = null, vararg params: Any?): T? {
   return this.invoke(target, *params) as? T?
 }

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/tasks/AndroidJUnit5JacocoReport.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/tasks/AndroidJUnit5JacocoReport.kt
@@ -2,10 +2,7 @@ package de.mannodermaus.gradle.plugins.junit5.tasks
 
 import com.android.build.gradle.api.BaseVariant
 import de.mannodermaus.gradle.plugins.junit5.*
-import de.mannodermaus.gradle.plugins.junit5.internal.android
-import de.mannodermaus.gradle.plugins.junit5.internal.extensionByName
-import de.mannodermaus.gradle.plugins.junit5.internal.junit5Info
-import de.mannodermaus.gradle.plugins.junit5.internal.maybeCreate
+import de.mannodermaus.gradle.plugins.junit5.internal.*
 import de.mannodermaus.gradle.plugins.junit5.providers.DirectoryProvider
 import de.mannodermaus.gradle.plugins.junit5.providers.mainClassDirectories
 import de.mannodermaus.gradle.plugins.junit5.providers.mainSourceDirectories
@@ -54,9 +51,7 @@ open class AndroidJUnit5JacocoReport : JacocoReport() {
       private val directoryProviders: Collection<DirectoryProvider>
   ) {
 
-    private val scope = variant.variantData.scope
-
-    val name: String = scope.getTaskName(TASK_NAME_DEFAULT)
+    val name: String = variant.getTaskName(prefix = TASK_NAME_DEFAULT)
 
     val type = AndroidJUnit5JacocoReport::class.java
 

--- a/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/tasks/AndroidJUnit5WriteFilters.kt
+++ b/plugin/android-junit5/src/main/kotlin/de/mannodermaus/gradle/plugins/junit5/tasks/AndroidJUnit5WriteFilters.kt
@@ -2,8 +2,8 @@ package de.mannodermaus.gradle.plugins.junit5.tasks
 
 import com.android.build.gradle.api.TestVariant
 import de.mannodermaus.gradle.plugins.junit5.INSTRUMENTATION_FILTER_RES_FILE_NAME
+import de.mannodermaus.gradle.plugins.junit5.internal.getTaskName
 import de.mannodermaus.gradle.plugins.junit5.internal.junit5ConfigurationOf
-import de.mannodermaus.gradle.plugins.junit5.variantData
 import org.gradle.api.DefaultTask
 import org.gradle.api.Project
 import org.gradle.api.tasks.CacheableTask
@@ -80,9 +80,7 @@ open class AndroidJUnit5WriteFilters : DefaultTask() {
       private val instrumentationTestVariant: TestVariant
   ) {
 
-    private val scope = instrumentationTestVariant.variantData.scope
-
-    val name: String = scope.getTaskName(TASK_NAME_DEFAULT)
+    val name: String = instrumentationTestVariant.getTaskName(prefix = TASK_NAME_DEFAULT)
 
     val type = AndroidJUnit5WriteFilters::class.java
 

--- a/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/FunctionalTests.kt
+++ b/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/FunctionalTests.kt
@@ -1,5 +1,6 @@
 package de.mannodermaus.gradle.plugins.junit5
 
+import de.mannodermaus.gradle.plugins.junit5.annotations.DisabledOnCI
 import de.mannodermaus.gradle.plugins.junit5.util.AgpVersion
 import de.mannodermaus.gradle.plugins.junit5.util.assertThat
 import de.mannodermaus.gradle.plugins.junit5.util.withPrunedPluginClasspath
@@ -14,9 +15,13 @@ class FunctionalTests {
 
   // Iterate over all values in the AgpVersion enum and,
   // using the project with that name located in the test resource folder,
-  // run a basic integration test with the JUnit 5 plugin
+  // run a basic integration test with the JUnit 5 plugin.
+  //
+  // The CI server does not have enough memory to run multiple nested virtual Gradle builds.
+  // Restrict execution of this test method to the local machine only
   @DisplayName("Integration Tests with Android Gradle Plugin")
   @TestFactory
+  @DisabledOnCI
   fun agpIntegrationTests(): List<DynamicTest> =
       AgpVersion.values()
           .map { agpVersion ->
@@ -38,46 +43,11 @@ class FunctionalTests {
               // Assert outputs;
               assertThat(result).task(":test").hasOutcome(TaskOutcome.SUCCESS)
 
-              // based on the project setup, assert different things.
-              // The assignment to the variable is unused, but required in order to enforce
-              // a case for each version
               with(result) {
-                @Suppress("UNUSED_VARIABLE")
-                val u: Any = when (agpVersion) {
-                  // ------------------------------------------------------------------------------------------------
-                  // AGP 3.5
-                  // - Product Flavors
-                  // - Build Types
-                  // ------------------------------------------------------------------------------------------------
-                  AgpVersion.AGP_35X -> {
-                    assertAgpTests(buildType = "debug", productFlavor = "free", tests = listOf("JavaTest"))
-                    assertAgpTests(buildType = "debug", productFlavor = "paid", tests = listOf("JavaTest", "KotlinPaidDebugTest"))
-                    assertAgpTests(buildType = "release", productFlavor = "free", tests = listOf("JavaTest", "KotlinReleaseTest", "JavaFreeReleaseTest"))
-                    assertAgpTests(buildType = "release", productFlavor = "paid", tests = listOf("JavaTest", "KotlinReleaseTest"))
-                  }
-                  // ------------------------------------------------------------------------------------------------
-                  // AGP 3.6
-                  // - Product Flavors
-                  // - Build Types
-                  // ------------------------------------------------------------------------------------------------
-                  AgpVersion.AGP_36X -> {
-                    assertAgpTests(buildType = "debug", productFlavor = "free", tests = listOf("JavaTest"))
-                    assertAgpTests(buildType = "debug", productFlavor = "paid", tests = listOf("JavaTest", "KotlinPaidDebugTest"))
-                    assertAgpTests(buildType = "release", productFlavor = "free", tests = listOf("JavaTest", "KotlinReleaseTest", "JavaFreeReleaseTest"))
-                    assertAgpTests(buildType = "release", productFlavor = "paid", tests = listOf("JavaTest", "KotlinReleaseTest"))
-                  }
-                  // ------------------------------------------------------------------------------------------------
-                  // AGP 4.0
-                  // - Product Flavors
-                  // - Build Types
-                  // ------------------------------------------------------------------------------------------------
-                  AgpVersion.AGP_40X -> {
-                    assertAgpTests(buildType = "debug", productFlavor = "free", tests = listOf("JavaTest"))
-                    assertAgpTests(buildType = "debug", productFlavor = "paid", tests = listOf("JavaTest", "KotlinPaidDebugTest"))
-                    assertAgpTests(buildType = "release", productFlavor = "free", tests = listOf("JavaTest", "KotlinReleaseTest", "JavaFreeReleaseTest"))
-                    assertAgpTests(buildType = "release", productFlavor = "paid", tests = listOf("JavaTest", "KotlinReleaseTest"))
-                  }
-                }
+                assertAgpTests(buildType = "debug", productFlavor = "free", tests = listOf("JavaTest"))
+                assertAgpTests(buildType = "debug", productFlavor = "paid", tests = listOf("JavaTest", "KotlinPaidDebugTest"))
+                assertAgpTests(buildType = "release", productFlavor = "free", tests = listOf("JavaTest", "KotlinReleaseTest", "JavaFreeReleaseTest"))
+                assertAgpTests(buildType = "release", productFlavor = "paid", tests = listOf("JavaTest", "KotlinReleaseTest"))
               }
             }
           }

--- a/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/annotations/DisabledOnCI.kt
+++ b/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/annotations/DisabledOnCI.kt
@@ -1,2 +1,6 @@
 package de.mannodermaus.gradle.plugins.junit5.annotations
 
+import org.junit.jupiter.api.condition.DisabledIfEnvironmentVariable
+
+@DisabledIfEnvironmentVariable(named = "CI", matches = "true")
+annotation class DisabledOnCI

--- a/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/annotations/DisabledOnCI.kt
+++ b/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/annotations/DisabledOnCI.kt
@@ -1,0 +1,2 @@
+package de.mannodermaus.gradle.plugins.junit5.annotations
+

--- a/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/util/AgpVersion.kt
+++ b/plugin/android-junit5/src/test/kotlin/de/mannodermaus/gradle/plugins/junit5/util/AgpVersion.kt
@@ -5,7 +5,8 @@ enum class AgpVersion(val fileKey: String,
 
   AGP_35X("agp35x"),
   AGP_36X("agp36x"),
-  AGP_40X("agp40x");
+  AGP_40X("agp40x"),
+  AGP_41X("agp41x", requiresGradle = "6.2");
 
   // Create a pretty string from the fileKey property.
   // Example:

--- a/plugin/android-junit5/src/test/projects/agp41x/build.gradle
+++ b/plugin/android-junit5/src/test/projects/agp41x/build.gradle
@@ -1,0 +1,47 @@
+buildscript {
+  repositories {
+    google()
+    jcenter()
+  }
+}
+
+plugins {
+  id "com.android.application"
+  id "org.jetbrains.kotlin.android"
+  id "de.mannodermaus.android-junit5"
+  id "jacoco"
+}
+
+def version = com.android.builder.model.Version.ANDROID_GRADLE_PLUGIN_VERSION
+if (version != AGP_41X) {
+  throw new IllegalStateException("Incorrect AGP version. Expected $AGP_41X, got $version")
+}
+
+repositories {
+  google()
+  jcenter()
+}
+
+android {
+  compileSdkVersion COMPILE_SDK_VERSION
+
+  flavorDimensions "environment"
+  productFlavors {
+    free {
+      dimension "environment"
+    }
+    paid {
+      dimension "environment"
+    }
+  }
+
+  testOptions.unitTests.all {
+    testLogging.events = ["passed", "skipped", "failed"]
+  }
+}
+
+dependencies {
+  implementation KOTLIN_STD_LIB
+  testImplementation JUPITER_API
+  testRuntimeOnly JUPITER_ENGINE
+}

--- a/plugin/android-junit5/src/test/projects/agp41x/settings.gradle
+++ b/plugin/android-junit5/src/test/projects/agp41x/settings.gradle
@@ -1,0 +1,15 @@
+pluginManagement {
+  repositories {
+    google()
+  }
+  resolutionStrategy {
+    eachPlugin {
+      if (requested.id.id == "com.android.application") {
+        useModule("com.android.tools.build:gradle:$AGP_41X")
+      }
+      if (requested.id.id == "org.jetbrains.kotlin.android") {
+        useModule("org.jetbrains.kotlin:kotlin-gradle-plugin:$KOTLIN")
+      }
+    }
+  }
+}

--- a/plugin/android-junit5/src/test/projects/agp41x/src/main/AndroidManifest.xml
+++ b/plugin/android-junit5/src/test/projects/agp41x/src/main/AndroidManifest.xml
@@ -1,0 +1,1 @@
+<manifest package="de.mannodermaus.app"></manifest>

--- a/plugin/android-junit5/src/test/projects/agp41x/src/main/java/de/mannodermaus/app/Adder.java
+++ b/plugin/android-junit5/src/test/projects/agp41x/src/main/java/de/mannodermaus/app/Adder.java
@@ -1,0 +1,7 @@
+package de.mannodermaus.app;
+
+public class Adder {
+  public int add(int a, int b) {
+    return a + b;
+  }
+}

--- a/plugin/android-junit5/src/test/projects/agp41x/src/test/java/de/mannodermaus/app/JavaTest.java
+++ b/plugin/android-junit5/src/test/projects/agp41x/src/test/java/de/mannodermaus/app/JavaTest.java
@@ -1,0 +1,13 @@
+package de.mannodermaus.app;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+class JavaTest {
+  @Test
+  void test() {
+    Adder adder = new Adder();
+    assertEquals(4, adder.add(2, 2), "This should succeed!");
+  }
+}

--- a/plugin/android-junit5/src/test/projects/agp41x/src/testFreeRelease/java/de/mannodermaus/app/JavaFreeReleaseTest.java
+++ b/plugin/android-junit5/src/test/projects/agp41x/src/testFreeRelease/java/de/mannodermaus/app/JavaFreeReleaseTest.java
@@ -1,0 +1,13 @@
+package de.mannodermaus.app;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+class JavaFreeReleaseTest {
+  @Test
+  void test() {
+    Adder adder = new Adder();
+    assertEquals(4, adder.add(2, 2), "This should succeed!");
+  }
+}

--- a/plugin/android-junit5/src/test/projects/agp41x/src/testPaidDebug/java/de/mannodermaus/app/KotlinPaidDebugTest.kt
+++ b/plugin/android-junit5/src/test/projects/agp41x/src/testPaidDebug/java/de/mannodermaus/app/KotlinPaidDebugTest.kt
@@ -1,0 +1,12 @@
+package de.mannodermaus.app
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+class KotlinPaidDebugTest {
+  @Test
+  fun test() {
+    val adder = Adder()
+    assertEquals(4, adder.add(2, 2), "This should succeed!")
+  }
+}

--- a/plugin/android-junit5/src/test/projects/agp41x/src/testRelease/java/de/mannodermaus/app/KotlinReleaseTest.kt
+++ b/plugin/android-junit5/src/test/projects/agp41x/src/testRelease/java/de/mannodermaus/app/KotlinReleaseTest.kt
@@ -1,0 +1,12 @@
+package de.mannodermaus.app
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+class KotlinReleaseTest {
+  @Test
+  fun test() {
+    val adder = Adder()
+    assertEquals(4, adder.add(2, 2), "This should succeed!")
+  }
+}

